File: data structures and algorithms/python/08_034.py
def trap_rainwater(heights):

    if not heights or len(heights) < 3:
        return 0
    
    left, right = 0, len(heights) - 1
    left_max, right_max = heights[left], heights[right]
    total_water = 0
    
    while left < right:
        if heights[left] < heights[right]:
            left += 1
            if heights[left] < left_max:
                total_water += left_max - heights[left]
            else:
                left_max = heights[left]
        else:
            right -= 1
            if heights[right] < right_max:
                total_water += right_max - heights[right]
            else:
                right_max = heights[right]
                
    return total_water

File: data structures and algorithms/cs/08_034.cs
public class RainwaterTrapper {
    public static int TrapRainwater(int[] heights) {
        if (heights == null || heights.Length < 3) {
            return 0;
        }

        int left = 0, right = heights.Length - 1;
        int leftMax = heights[left], rightMax = heights[right];
        int totalWater = 0;

        while (left < right) {
            if (heights[left] < heights[right]) {
                left++;
                if (heights[left] < leftMax) {
                    totalWater += leftMax - heights[left];
                } else {
                    leftMax = heights[left];
                }
            } else {
                right--;
                if (heights[right] < rightMax) {
                    totalWater += rightMax - heights[right];
                } else {
                    rightMax = heights[right];
                }
            }
        }

        return totalWater;
    }
}